test/parallel_api/numeric/numeric.ops/inclusive_scan_by_segment.pass.cpp

```diff
@@ -16,98 +16,116 @@
-        //T vals[n1] = { 1, 1, 1, ... };
+        //T keys[n] = { 1, 2, 3, 4, 1, 1, 2, 2, 3, 3, 4, 4, 1, 1, 1, ...};
+        //T vals[n] = { n random numbers between 0 and 4 };
 
+        ::std::srand(42);
         Size segment_length = 1;
-        for (Size i = 0; i != n; )
+        for (Size i = 0; i != n;)
         {
-          for (Size j = 0; j != 4*segment_length && i != n; ++j)
-          {
-              host_keys[i] = j/segment_length + 1;
-              host_vals[i] = 1;
-              host_val_res[i] = 0;
-              ++i;
-          }
-          ++segment_length;
+            for (Size j = 0; j != 4 * segment_length && i != n; ++j)
+            {
+                host_keys[i] = j / segment_length + 1;
+                host_vals[i] = rand() % 5;
+                host_val_res[i] = 0;
+                ++i;
+            }
+            ++segment_length;
         }
     }
 
 #ifdef DUMP_CHECK_RESULTS
     template <typename Iterator, typename Size>
     void display_param(const char* msg, Iterator it, Size n)
     {
-        std::cout << msg;
+        ::std::cout << msg;
         for (Size i = 0; i < n; ++i)
         {
             if (i > 0)
-                std::cout << ", ";
-            std::cout << it[i];
+                ::std::cout << ", ";
+            ::std::cout << it[i];
         }
-        std::cout << std::endl;
+        ::std::cout << ::std::endl;
     }
 #endif // DUMP_CHECK_RESULTS
 
     template <typename Iterator1, typename Iterator2, typename Iterator3, typename Size,
+              typename BinaryPredicateCheck = oneapi::dpl::__internal::__pstl_equal,
               typename BinaryOperationCheck = oneapi::dpl::__internal::__pstl_plus>
-    void
-    check_values(Iterator1 host_keys, Iterator2 host_vals, Iterator3 val_res, Size n,
-                 BinaryOperationCheck op = BinaryOperationCheck())
+    void check_values(Iterator1 host_keys, Iterator2 host_vals, Iterator3 val_res, Size n,
+                      BinaryPredicateCheck pred = BinaryPredicateCheck(),
+                      BinaryOperationCheck op = BinaryOperationCheck())
     {
         // https://docs.oneapi.io/versions/latest/onedpl/extension_api.html
         // keys:   [ 0, 0, 0, 1, 1, 1 ]
         // values: [ 1, 2, 3, 4, 5, 6 ]
         // result: [ 1, 1 + 2 = 3, 1 + 2 + 3 = 6, 4, 4 + 5 = 9, 4 + 5 + 6 = 15 ]
 
-#ifdef DUMP_CHECK_RESULTS
-        std::cout << "check_values(n = " << n << ") : " << std::endl;
-        display_param("keys:   ", host_keys, n);
-        display_param("values: ", host_vals, n);
-        display_param("result: ", val_res,   n);
-#endif // DUMP_CHECK_RESULTS
-
         if (n < 1)
             return;
 
-        using ValT = typename ::std::decay<decltype(val_res[0])>::type;
+        typedef typename ::std::iterator_traits<Iterator1>::value_type KeyT;
+        typedef typename ::std::iterator_traits<Iterator2>::value_type ValT;
+
+        ::std::string failure_msg = "wrong effect from inclusive_scan_by_segment with types [";
```

Comment:
may be better to use ```std::string_stream```?

Reply:
Yes, I agree since we are concatenating many strings. I used `std::ostringstream`.

Reply:
Do you mean ```std::stringstream``` ?

Reply:
Since we are using it for output only, I used [std::ostringstream](https://en.cppreference.com/w/cpp/io/basic_ostringstream) which is used in a few other places. It provides the same functionality as `std::stringstream` but only for output.

Reply:
OK